migration-engine/core/src/state.rs

```diff
@@ -0,0 +1,270 @@
+use crate::{
+    api::GenericApi,
+    commands,
+    json_rpc::types::{
+        ApplyMigrationsInput, ApplyMigrationsOutput, CreateDatabaseParams, CreateMigrationInput, CreateMigrationOutput,
+        DatasourceParam, DbExecuteDatasourceType, DbExecuteParams, DevDiagnosticInput, DevDiagnosticOutput, DiffParams,
+        DiffResult, EnsureConnectionValidityParams, EvaluateDataLossInput, EvaluateDataLossOutput,
+        ListMigrationDirectoriesInput, ListMigrationDirectoriesOutput, MarkMigrationAppliedInput,
+        MarkMigrationAppliedOutput, MarkMigrationRolledBackInput, MarkMigrationRolledBackOutput, PathContainer,
+        SchemaContainer, SchemaPushInput, SchemaPushOutput, UrlContainer,
+    },
+    CoreResult,
+};
+use enumflags2::BitFlag;
+use migration_connector::{ConnectorError, ConnectorHost, MigrationConnector};
+use std::{collections::HashMap, future::Future, path::Path, pin::Pin, sync::Arc};
+use tokio::sync::Mutex;
+use tracing_futures::Instrument;
+
+/// The container for the state of the migration engine.
+pub(crate) struct EngineState {
+    initial_datamodel: Option<String>,
+    host: Arc<dyn ConnectorHost>,
+    connectors: Mutex<HashMap<String, Box<dyn MigrationConnector>>>,
```

Comment:
document: the string is

- the complete schema string
- or the database URL, for connections/connectors that don't come from a schema